Zend/zend_property_hooks.c

```diff
@@ -252,14 +254,11 @@ static void zho_it_move_forward(zend_object_iterator *iter)
 		if (zend_hash_has_more_elements(properties) != SUCCESS) {
 			hooked_iter->declared_props_done = true;
```

Comment:
Can we also move this check to stay consistent between declared and dynamic props?

Reply:
I've tried, but it's not trivial. If `zho_it_move_forward()` doesn't mark `declared_props_done` immediately we will call `zho_declared_it_fetch_current()`, and `zho_it_move_forward()` again. But this time we will increment the dynamic position and skip the first dynamic property.

We would also get similar issues if the client called `->move_forward()` a few times without calling `->get_current()`.

We could simplify how we chose between declared and dynamic props. E.g. use declared ones when `declared_props.nInternalPointer < declared_props.nNumUsed`, and dynamic ones otherwise. We can then remove `declared_props_done` / `dynamic_props_done`:

``` diff
diff --git a/Zend/zend_property_hooks.c b/Zend/zend_property_hooks.c
index d701abdc3c5..c9d52f18c9f 100644
--- a/Zend/zend_property_hooks.c
+++ b/Zend/zend_property_hooks.c
@@ -25,9 +25,7 @@
 typedef struct {
 	zend_object_iterator it;
 	bool by_ref;
-	bool declared_props_done;
 	zval declared_props;
-	bool dynamic_props_done;
 	uint32_t dynamic_prop_it;
 	zval current_key;
 	zval current_data;
@@ -100,7 +98,6 @@ static void zho_dynamic_it_init(zend_hooked_object_iterator *hooked_iter)
 {
 	zend_object *zobj = Z_OBJ_P(&hooked_iter->it.data);
 	zend_array *properties = zobj->handlers->get_properties(zobj);
-	hooked_iter->dynamic_props_done = false;
 	hooked_iter->dynamic_prop_it = zend_hash_iterator_add(properties, zho_num_backed_props(zobj));
 }
 
@@ -161,7 +158,6 @@ static void zho_dynamic_it_fetch_current(zend_object_iterator *iter)
 	HashPosition pos = zend_hash_iterator_pos(hooked_iter->dynamic_prop_it, properties);
 
 	if (pos >= properties->nNumUsed) {
-		hooked_iter->dynamic_props_done = true;
 		return;
 	}
 
@@ -191,10 +187,12 @@ static void zho_it_fetch_current(zend_object_iterator *iter)
 		return;
 	}
 
+	zend_array *properties = Z_ARR(hooked_iter->declared_props);
+
 	while (true) {
-		if (!hooked_iter->declared_props_done) {
+		if (properties->nInternalPointer < properties->nNumUsed) {
 			zho_declared_it_fetch_current(iter);
-		} else if (!hooked_iter->dynamic_props_done) {
+		} else if (zend_hash_iterator_pos(hooked_iter->dynamic_prop_it, Z_OBJ(iter->data)->properties) < Z_OBJ(iter->data)->properties->nNumUsed) {
 			zho_dynamic_it_fetch_current(iter);
 		} else {
 			break;
@@ -248,13 +246,10 @@ static void zho_it_move_forward(zend_object_iterator *iter)
 	zval_ptr_dtor_nogc(&hooked_iter->current_key);
 	ZVAL_UNDEF(&hooked_iter->current_key);
 
-	if (!hooked_iter->declared_props_done) {
-		zend_array *properties = Z_ARR(hooked_iter->declared_props);
-		zend_hash_move_forward(properties);
-		if (zend_hash_has_more_elements(properties) != SUCCESS) {
-			hooked_iter->declared_props_done = true;
-		}
-	} else if (!hooked_iter->dynamic_props_done) {
+	zend_array *properties = Z_ARR(hooked_iter->declared_props);
+	if (properties->nInternalPointer < properties->nNumUsed) {
+		properties->nInternalPointer++;
+	} else {
 		zend_array *properties = Z_OBJ(iter->data)->properties;
 		HashPosition pos = zend_hash_iterator_pos(hooked_iter->dynamic_prop_it, properties);
 		pos++;
@@ -271,13 +266,9 @@ static void zho_it_rewind(zend_object_iterator *iter)
 	zval_ptr_dtor_nogc(&hooked_iter->current_key);
 	ZVAL_UNDEF(&hooked_iter->current_key);
 
-	hooked_iter->declared_props_done = false;
 	zend_array *properties = Z_ARR(hooked_iter->declared_props);
 	zend_hash_internal_pointer_reset(properties);
-	hooked_iter->dynamic_props_done = false;
-	if (hooked_iter->dynamic_prop_it != (uint32_t) -1) {
-		EG(ht_iterators)[hooked_iter->dynamic_prop_it].pos = zho_num_backed_props(Z_OBJ(iter->data));
-	}
+	EG(ht_iterators)[hooked_iter->dynamic_prop_it].pos = zho_num_backed_props(Z_OBJ(iter->data));
 }
 
 static HashTable *zho_it_get_gc(zend_object_iterator *iter, zval **table, int *n)
@@ -318,7 +309,6 @@ ZEND_API zend_object_iterator *zend_hooked_object_get_iterator(zend_class_entry
 	ZVAL_OBJ_COPY(&iterator->it.data, zobj);
 	iterator->it.funcs = &zend_hooked_object_it_funcs;
 	iterator->by_ref = by_ref;
-	iterator->declared_props_done = false;
 	zend_array *properties = zho_build_properties_ex(zobj, true, false);
 	ZVAL_ARR(&iterator->declared_props, properties);
 	zho_dynamic_it_init(iterator);
```

WDYT? This may be for an other PR however.